src/jit/MemoryInl.h

```diff
@@ -1390,15 +1391,70 @@ static void emitAtomic(sljit_compiler* compiler, Instruction* instr)
         sljit_s32 baseReg = SLJIT_EXTRACT_REG(addr.memArg.arg);
         sljit_s32 tmpReg = srcReg;
 
+        JITArg memValue(operands + 0);
+        sljit_s32 memValueReg = SLJIT_EXTRACT_REG(memValue.arg);
+        sljit_s32 maskReg;
+        sljit_s32 shiftReg;
+
+        if (noShortAtomic && size <= 2) {
+            operationSize = SLJIT_MOV32;
+
+            shiftReg = instr->requiredReg(2);
+            maskReg = instr->requiredReg(3);
+        }
+
+        if (SLJIT_IS_IMM(memValueReg)) {
```

Comment:
What is this case?

Reply:
In the case of unshared memory, the arguments are immediate zeroes
https://github.com/Samsung/walrus/blob/7f492b3ffc7361b94b32085117ecde1915565eea/test/extended/threads/atomic.wast#L440

Reply:
Upon further investigation `memValueReg` is only an immediate in the RMW cases (not RMW Cmpxchg), but in my observation only in the unshared memory test cases.
But the deciding factor is not the memory being unshared, as the code works fine without the check for immediates even if I remove the shared flag from the very first module at the top of the test file for example.

However just returning in these cases is most likely incorrect; I have updated it so a TMP register is assigned in these cases instead.